src/cache_cf.cc

```diff
@@ -550,10 +543,17 @@ parseOneConfigFile(const char *file_name, unsigned int depth)
                         debugs(3, DBG_CRITICAL, ConfigParser::CurrentLocation() << ": unrecognized: '" << tmp_line << "'");
                         ++err_count;
                     }
+                } catch (const Cfg::FatalError &e) {
+                    debugs(3, DBG_CRITICAL, e.what());
+                    // when -k parse is used try to report as many admin mistakes as possible before terminating
+                    if (opt_parse_cfg_only)
+                        ++err_count;
+                    else
+                        LegacyParser.destruct();
```

Comment:
I agree that this is an extra feature that it makes sense to disentangle from the self_destruct removal.
I like it tho, and I hope it can be reintroduced in a followup PR.

Reply:
This is how -k parse works differently from -k check. The parse-only mode is verbose about errors and exits only at the end of the parse. Whereas -k check is silent unless 1 error occurs, in which case it destructs immediately.

The only thing changing in this PR is that the the keep-going is being done centrally here instead of ad-hoc all over the code. The ad-hoc tend to be conditional fatal() calls rather than self_destruct, so not showing up very clearly in this PR change. Cleaning those to use this new exception type is left for future work.

Reply:
> This is how -k parse works differently from -k check. The parse-only mode ... exits only at the end of the parse. Whereas -k check ... destructs [on the first error].

AFAICT, you are implying that `-k parse` plows through all errors to the end of configuration while `-k check` stops on the first error. While that ideal makes sense, in reality, `-k parse` may quit on the first error as well, ignoring all other remaining errors. It all depends on the error the code encounters.

Moreover, supporting that "to report all errors, always keep going" ideal is impractical and probably impossible. Even compilers cannot do that. There will always be two kinds of errors -- some should terminate parsing (because searching for more errors is likely to be counterproductive) and some should be bypassed (e.g., because they are likely to have small/local impact).

This PR has to make a choice:

1. Confirm that, in this PR, we are only interested in improving handling of previously fatal errors -- the code that used to call self_destruct() and such. Throw one exception type. Quit ASAP after that caught exception is reported. The `-k` mode remains irrelevant.
2. Enlarge PR scope to also improve handling of some bypassable errors. Throw two kinds of exceptions (either two types or, more likely, one type with a `bypassable` data member).

I am OK with taking one or both steps. It is your call. There is also step 3 that deals with reconfigurations, but we are probably too far from that step to discuss it here.

If you decide to limit this PR to step 1, the existence of future step 2 needs may still affect our step 1 design choices, but that effect would be very minor. For example, we may decide not to call the thrown exception _type_ FatalError. The PR would still throw the result of `FatalError(...)` though. FatalError() will just be an (inlined) function returning a newly created Cfg::Error, not a constructor.

Do you want to limit this PR to step 1?

Reply:
> > This is how -k parse works differently from -k check. The parse-only mode ... exits only at the end of the parse. Whereas -k check ... destructs [on the first error].
> 
> AFAICT, you are implying that `-k parse` plows through all errors to the end of configuration while `-k check` stops on the first error. While that ideal makes sense, in reality, `-k parse` may quit on the first error as well, ignoring all other remaining errors. It all depends on the error the code encounters.
> 

I do not mean to imply, I mean to outright state that behaviour as one of the designer intended differences between parse and check. The fact that some code is/was aborting immediately in the parse case when it could have continued processing *other* lines is a bug.

In all modes except "parse" the -k should be loading the config (abort on first error) then signalling the running process to do whatever action. That includes "check".

The purpose of "parse" being a different command from "check" is to allow for auditing the config file without involving any updates to the running process. There is very low utility in an auditor that only reports one problem at a time.


The if-statement this thread is attached to was done in this PR because:
 a) some of the self_destruct() calls being removed were conditional on this same statement (in their old code paths), and
 b) the other self_destruct() calls were buggy by not having this conditional, and
 c) it cannot be added first because the exception does not exist until this PR, and
 d) delaying it to a later PR actively turns the code paths from (a) into even more bugs.




> > This is how -k parse works differently from -k check. The parse-only mode ... exits only at the end of the parse. Whereas -k check ... destructs [on the first error].
> 
> AFAICT, you are implying that `-k parse` plows through all errors to the end of configuration while `-k check` stops on the first error. While that ideal makes sense, in reality, `-k parse` may quit on the first error as well, ignoring all other remaining errors. It all depends on the error the code encounters.
> 
> Moreover, supporting that "to report all errors, always keep going" ideal is impractical and probably impossible. Even compilers cannot do that. There will always be two kinds of errors -- some should terminate parsing (because searching for more errors is likely to be counterproductive) and some should be bypassed (e.g., because they are likely to have small/local impact).
> 

I think you are missing part of what is implemented here. There *are* groups of errors as you describe. In fact there are four:

 1) Cfg::FatalError() - halts processing the current squid.conf line. Any subsequent processing of that line is abandoned as that would compound problems.
 --> When "k parse" is used *other* lines continue processing before the final exit aborts Squid.
 
 2) TexcHere() - halt whatever AsyncJob / activity was happening. Cfg uses this for ignorable errors with the throw as fallback or admin configured action. The active process continues with any other activities unrelated to config parsing.
 --> When "k reconfigure" is used active traffic may continue with old config settings (or not).
  [ modulo the bug of initially parsed settings being used after the exception. ]

3) Must()  - halt immediately with an error message and state cleanup.

4) assert() - halt immediately with no useful output or state alteration.


> This PR has to make a choice:
> 
>     1. Confirm that, in this PR, we are only interested in improving handling of previously fatal errors -- the code that used to call self_destruct() and such. Throw one exception type. Quit ASAP after that caught exception is reported. The `-k` mode remains irrelevant.
> 
>     2. Enlarge PR scope to also improve handling of some bypassable errors. Throw two kinds of exceptions (either two types or, more likely, one type with a `bypassable` data member).
> 

Those are not the only options. I chose a third far simpler one when scoping this PRs work in the first place:

3) To implement only replacing self_destruct() with an exception. Moving the code around those self_destruct() to the relevant centralized catch statement(s).
 
De-duplicating some very obviously repeated code with those Cfg::Require*() gadgets was a slight extension. Though differences from existing code minimized as much as possible.

As mentioned already, this controversial if-statement is here because it was _existing behaviour_, existing code even.

Reply:
I think we are largely in agreement here. I see problems in details, but after writing a lengthy response I removed it -- nothing you have said overall contradicts my understanding or addresses this change request. I suspect/hope that there is just some high-level misunderstanding and hashing out those low-level problems is not going to help.

It is not yet clear to me what exactly you think is wrong in my change request. In hope to make progress, I will share my plan for going forward:

Now that I know that this PR wants to make _all_ previously unconditional self_destruct() errors "non-final" (for the lack of a better word), I will re-review the PR:

* If we can do this change safely, then we can use just one new error kind. My original change request will still apply and can be easily addressed by renaming the error we are throwing. In this happy case, `squid -k` will continue parsing after catching that non-final error.

* Otherwise (i.e. if we find cases where proceeding with parsing after self_destruct() is unsafe), we will need two error kinds to fully address this change request. `squid -k` will continue parsing after catching the non-final error kind.

In either case, we will need to adjust the PR description to disclose `squid -k` changes because Squid will be able to report more errors than before (for some broken configurations).

Naturally, reviewing 200+ self_destruct() calls will take some time. Please bear with me.